### PR TITLE
Add sign up page to image sharer

### PR DESCRIPTION
**Pull request by mathewbaltes**

### CODE REVIEW

File: test/controllers/users_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    assert_difference 'User.count', 1 do
+      post users_path, params: {
+        user: { name: 'a', email: 'a@a.com', password: 'blah123', password_confirmation: 'blah123' }
+      }
+    end
+
+    created_user = User.last
+
+    assert_redirected_to root_path
+    assert_equal 'a', created_user.name
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Is the controller test the right place to check for if the attributes are stored correctly?

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'create a valid user' do
+    assert_difference 'User.count', 1 do
+      post users_path, params: {
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Is there another HTTP response related side effect to check for?

File: test/controllers/users_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    assert_equal 'a@a.com', created_user.email
+  end
+
+  test 'create an invalid user' do
+    assert_no_difference 'User.count' do
+      post users_path, params: {
+        user: { name: 'a', email: 'a@a.com', password: 'blah123', password_confirmation: '123' }
+      }
+    end
+    assert_response :ok
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Why ok?

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @user = User.new(
+      name: 'Example User',
+      email: 'user@app.com',
+      password: '123456',
+      password_confirmation: '123456'
+    )
+  end
+
+  test 'should be valid' do
+    assert @user.valid?
```
Comment (bboe) [inline test/models/user_test.rb]: Consider using `assert_predicate @user, :valid?`. Compare the error messages when they fail to see why one is preferred.

If you must use `assert`, please include a descriptive error message as the second argument.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  test 'name should be present' do
+    @user.name = ''
+    assert_not @user.valid?
```
Comment (bboe) [inline test/models/user_test.rb]: @user has an `errors` attribute that is set when it's not valid. Perhaps you can make some assertions in all the failure cases for how errors is set to ensure it fails for the reason you are testing for.

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    duplicate_user = @user.dup
+    duplicate_user.email = @user.email.upcase
+    @user.save
+    assert_not duplicate_user.valid?
+  end
+
+  test 'email addresses should be lowercase' do
+    @user.email = @user.email.upcase
+    @user.save
+    User.last.email = @user.email.downcase
```
Comment (bboe) [inline test/models/user_test.rb]: What's the purpose of this statement?

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  test 'password with less than six characters is invalid' do
```
Comment (bboe) [inline test/models/user_test.rb]: Perhaps you should test the edge case, 5 to ensure that it also fails there. It's always a good idea to enforce that your constraints do not easily change by adding test cases around those constraints.

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<%= simple_form_for @user, url: users_path do |f| %>
+    <div class="row justify-content-center">
+      <div class="col-lg-6">
+        <%= f.input :name, label: 'Name', as: :string, input_html: { placeholder: 'Enter name' } %>
+        <%= f.input :email, label: 'Email', as: :string, input_html: { placeholder: 'Enter email' } %>
+        <%= f.input :password, label: 'Password', as: :string, input_html: { placeholder: 'Enter password' } %>
```
Comment (mathewbaltes) [inline app/views/users/new.html.erb]: as: :password, did this on the login but missed it here.

File: config/routes.rb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
-  resources :images, except: [:edit, :update]
+  resources :images, only: [:index, :create, :new, :show, :destroy]
```
Comment (mathewbaltes) [inline config/routes.rb]: No, I'll remove it.

File: test/models/user_test.rb (hunk 7 above)
Comment (mathewbaltes) [inline test/models/user_test.rb]: Agreed.

File: test/models/user_test.rb (hunk 6 above)
Comment (mathewbaltes) [inline test/models/user_test.rb]: Meant assert_equal, nice catch.

File: test/controllers/users_controller_test.rb (hunk 1 above)
Comment (mathewbaltes) [inline test/controllers/users_controller_test.rb]: No, we should be testing that functionality with the model.

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.valid?
+      @user.save!
+      flash[:success] = "Thank you for signing up #{@user.name}"
+      redirect_to root_path
+    else
+      render :new
```
Comment (mathewbaltes) [inline app/controllers/users_controller.rb]: Updated to 422.

File: db/migrate/20170413185550_create_users.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :name
```
Comment (bboe) [inline db/migrate/20170413185550_create_users.rb]: Should these be required?

**Comment by bboe** [comment]
Great work!

File: db/migrate/20170413185550_create_users.rb (hunk 11 above)
Comment (mathewbaltes) [inline db/migrate/20170413185550_create_users.rb]: Yes, probably should be required!
